File: src/service/mq/rpc.ts
import { MQBase } from './base';
import type * as amqplib from 'amqplib';
import { v4 } from 'uuid';

export async function EchoHandler(msg: string) {
    return msg;
}

class RPCBase extends MQBase {
    async assertServerQueue(queueName: string) {
        return await this.assertQueue(queueName, { durable: false })
    }

    async assertClientQueue() {
        return await this.assertQueue('', { exclusive: true });
    }
}


export class RPCServer extends RPCBase {
    queue: Promise<{ ch: amqplib.Channel, queue: string }>;
    handler: (msg: string) => Promise<string>;
    constructor(url: string, handler = EchoHandler) {
        super(url);
        this.handler = handler;
        this.queue = this.assertServerQueue('rpc_' + v4());
        this.run();
    }

    private async run() {
        const { ch, queue } = await this.queue;
        await ch.prefetch(1);
        ch.consume(queue, async (msg: amqplib.ConsumeMessage | null) => {
            if (!msg) return;
            ch.ack(msg);
            const { replyTo, correlationId } = msg.properties;
            const result = await this.handler(msg.content.toString());
            ch.sendToQueue(replyTo, Buffer.from(result), { correlationId });
        });
    }
}


export class RPCClient extends RPCBase {
    resultQueue: Promise<{ ch: amqplib.Channel, queue: string }>;
    resolveMap = new Map<string, (result: string) => void>();
    constructor(url: string) {
        super(url);
        this.resultQueue = this.assertClientQueue();
        this.run();
    }

    private async run() {
        const { ch, queue } = await this.resultQueue;
        ch.consume(queue, (msg: amqplib.ConsumeMessage | null) => {
            if (!msg) return;
            const resolve = this.resolveMap.get(msg.properties.correlationId);
            if (resolve) resolve(msg.content.toString());
            this.resolveMap.delete(msg.properties.correlationId);
        });
    }

    private getResult(correlateId: string, timeout: number = 100000) {
        return new Promise<string>((resolve, reject) => {
            const timeoutId = setTimeout(() => {
                reject(new Error('timeout'));
            }, timeout);
            this.resolveMap.set(correlateId, (msg: string) => {
                resolve(msg);
                clearTimeout(timeoutId);
            });
        });
    }

    async call(msg: string, sQueueName: string): Promise<any> {
        const { queue, ch } = await this.resultQueue;
        const sQueue = (await this.assertServerQueue(sQueueName)).queue;
        const correlationId = v4();
        await ch.sendToQueue(sQueue, Buffer.from(msg), { correlationId, replyTo: queue });
        return await this.getResult(correlationId);
    }
}

File: csharp/tmlib/MQ/RPC.cs
using RabbitMQ.Client.Events;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AniTask.MQ
{
  public class RPCBase : MQBase
  {
    public RPCBase(string url) : base(url) { }
    public string AssertServerQueue(string queue)
    {
      var queueOK = connection.ch.QueueDeclare(queue,
                           durable: false,
                           exclusive: false,
                           autoDelete: false,
                           arguments: null);

      return queueOK.QueueName;
    }

    public string AssertClientQueue()
    {
      return connection.ch.QueueDeclare("", durable: false,
                           exclusive: true,
                           autoDelete: false,
                           arguments: null).QueueName;
    }
  }


  public class RPCServer : RPCBase
  {
    public string queue { get; }
    public delegate string Handler(string msg);

    public RPCServer(string url) : base(url)
    {
      this.queue = this.AssertServerQueue("rpc_server_" + Utils.GUID());
    }

    public bool Run(Handler handler)
    {
      var channel = this.connection.ch;
      channel.BasicQos(0, 1, false);

      var consumer = new EventingBasicConsumer(channel);
      channel.BasicConsume(queue: this.queue,
        consumerTag: "",
        noLocal:false,
        exclusive:false,
        arguments:null,
        autoAck: false,
        consumer: consumer);
      Console.WriteLine(" [x] Awaiting RPC requests");
      consumer.Received += (model, ea) =>
      {
        string response = null;

        var body = ea.Body.ToArray();
        var props = ea.BasicProperties;
        var replyProps = channel.CreateBasicProperties();
        replyProps.CorrelationId = props.CorrelationId;
        var message = Encoding.UTF8.GetString(body);
        channel.BasicAck(deliveryTag: ea.DeliveryTag,
          multiple: false);
 
        var result = handler.Invoke(message);
        var responseBytes = Encoding.UTF8.GetBytes(response);
        channel.BasicPublish(exchange: "",
          routingKey: props.ReplyTo,
          mandatory: false,
          basicProperties: replyProps,
          body: responseBytes);
      };
      return true;
    }
  }


  public class RPCClient : RPCBase
  {
    protected string replyQueueName;
    private EventingBasicConsumer consumer;
    private readonly BlockingCollection<string> respQueue = new BlockingCollection<string>();
    string correlationId;
    public RPCClient(string url):base(url)
    {
      this.replyQueueName = AssertClientQueue();
      this.Run();
    }

    private void Run()
    {
      var channel = this.connection.ch;
      consumer = new EventingBasicConsumer(channel);
      consumer.Received += (model, ea) =>
      {
        var body = ea.Body.ToArray();
        var response = Encoding.UTF8.GetString(body);
        if (ea.BasicProperties.CorrelationId == correlationId)
        {
          respQueue.Add(response);
        }
      };
      channel.BasicConsume(
          consumer: consumer,
          queue: replyQueueName,
          consumerTag: "",
          noLocal: false,
          exclusive: false,
          arguments: null,
          autoAck: true);
    }

    public string Call(string message, string serverQueue)
    {
      correlationId = Guid.NewGuid().ToString();
      var channel = this.connection.ch;
      var props = channel.CreateBasicProperties();
      props.CorrelationId = correlationId;
      props.ReplyTo = replyQueueName;

      var messageBytes = Encoding.UTF8.GetBytes(message);
      channel.BasicPublish(
          exchange: "",
          routingKey: serverQueue,
          mandatory: false,
          basicProperties: props,
          body: messageBytes);
      return respQueue.Take();
    }
  }
}
